# KNNFewShot

Companion notebook for Chapter 6, "KNNFewShot" (pages 140–141).

Retrieve the training examples nearest to each input and compile a small few-shot program for it at inference time.

**Use it when:** Different inputs benefit from different demonstrations and a retriever over your training examples is available.

**What compilation changes:** The instruction stays fixed; the demonstrations are selected by similarity for each input.

Important configuration in this benchmark:

- `k=4` nearest training examples for each input
- `hashed_ngram_embeddings` builds local 512-dimensional hashed unigram/bigram vectors, so no embedding API is called
- `trainset` and `vectorizer` go to the constructor; `compile` takes only the program

Every notebook in this folder loads the chapter's 300-passage benchmark
and its frozen, pair-grouped split: 160 training, 60 validation, and 80
locked-test examples. Both members of a human/AI pair stay in the same
partition. Optimizer choices use validation only; the locked test is
scored once, after the program is frozen. These scores compare optimizer
behavior; they are not a general AI-detector leaderboard.

**Keys and cost:** The default **Run All** is free: it needs no API key, makes no model call, and shows the saved result of the recorded run. A live run (`CHAPTER06_RUN_LIVE=1`) makes paid model calls and needs `OPENAI_API_KEY`. Compiling made no model call in the recorded run; evaluating the compiled program cost $0.2380. A live run is a new experiment, so its score, cost, and time will differ.

In [1]:
import os
import sys
from pathlib import Path

import dspy
from dotenv import load_dotenv

cwd = Path.cwd().resolve()
REPO_ROOT = cwd if (cwd / "chapter06").is_dir() else cwd.parent
if not (REPO_ROOT / "chapter06" / "results" / "expanded_notebooks" / "comparison.json").exists():
    raise RuntimeError("Run this notebook from the repository root or the chapter06 directory.")
if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))
load_dotenv(REPO_ROOT / ".env")

from chapter06.notebook_support import artifact_paths, learned_program_preview, verify_prompt_artifact
from chapter06.optimizer_runtime import format_result, published_result, score_summary

OPTIMIZER = 'knn-few-shot'
RUN_LIVE = os.getenv("CHAPTER06_RUN_LIVE", "0") == "1"
if RUN_LIVE and not os.getenv("OPENAI_API_KEY"):
    raise RuntimeError(
        "CHAPTER06_RUN_LIVE=1 needs OPENAI_API_KEY. Add it to the .env file "
        "in the repository root and restart the kernel."
    )
print(f"optimizer={OPTIMIZER!r}; live={RUN_LIVE}")

optimizer='knn-few-shot'; live=False


## Load the frozen dataset

The next cell is the setup listing from Chapter 6, "Setting Up the AI Detection
Module" (page 135). Every notebook in this folder starts with it.

Cells that call a model API hold the chapter listing unchanged, indented under
`if RUN_LIVE:`. By default they are skipped, so **Run All** is free and works
offline, and the saved result of the recorded run is displayed instead. To run
the listings yourself, set `CHAPTER06_RUN_LIVE=1` before starting Jupyter.
Every live run needs `OPENAI_API_KEY` in the repository `.env` file; if the
key is missing, the setup cell above stops with a message.

In [2]:
# Load the frozen Chapter 6 dataset and pair-grouped split
from chapter06.optimizer_runtime import (
    AIDetector,
    exact_match,
    hashed_ngram_embeddings,
    load_frozen_examples,
    split_summary,
)

NUM_THREADS = 1
splits = load_frozen_examples()
trainset = splits["train"]
valset = splits["validation"]
testset = splits["test"]

print(split_summary(splits))

train=160 (human=80, AI=80); validation=60 (human=30, AI=30); test=80 (human=40, AI=40)


> **Book vs. notebook:** The chapter creates `task_lm` and configures DSPy in its baseline listing (page 135). The cell below repeats those two steps so this notebook runs on its own. Creating the model object makes no API call.

In [3]:
task_lm = dspy.LM("openai/gpt-5.6-luna", cache=False)
dspy.configure(lm=task_lm)
print(f"task model: {task_lm.model}")

task model: openai/gpt-5.6-luna


## Compile with KNNFewShot

Page 140. Building the optimizer embeds the training split locally and compiling makes no model call, so this cell runs in the default mode too.

In [4]:
# Set up the optimizer and training data
optimizer = dspy.teleprompt.KNNFewShot(
        k=4,
        metric=exact_match,
        trainset=trainset,
        vectorizer=dspy.Embedder(hashed_ngram_embeddings),

    )


optimized_detector = optimizer.compile(
    AIDetector()
)

## Score the compiled program

In live mode the next cell scores `optimized_detector`, the program the listing above produced, on the validation split and then once on the locked test. As printed, each prediction first bootstraps demonstrations on its four retrieved neighbors, so it makes about five model calls: roughly 700 calls and close to half an hour for the 60 validation and 80 locked-test predictions. DSPy prints a progress bar and a `Bootstrapped ...` line for every prediction; the summary lines come last.

A call that fails, or an answer that cannot be read as true or false, counts as a wrong prediction; the summary then ends with a `Failed Calls` line. If the first three calls fail outright, or no validation prediction can be scored, the cell stops and shows the first error.

In [5]:
if RUN_LIVE:
    print(score_summary(optimized_detector, valset, testset))
else:
    print(
        "Skipped: scoring calls the model. Set CHAPTER06_RUN_LIVE=1 before "
        "starting Jupyter to run it; the saved result is shown below."
    )

Skipped: scoring calls the model. Set CHAPTER06_RUN_LIVE=1 before starting Jupyter to run it; the saved result is shown below.


## Saved result

The block below is the recorded run for this notebook, in the layout of the result blocks printed in the chapter. Every notebook uses the same layout, so a block can show details the chapter prints for some optimizers only: the count behind the validation percentage, token totals, the combined cost, the time in minutes, and percentages to two decimal places. It is read from `results/expanded_notebooks/comparison.json`, so it is the same whether or not you run live. The recorded run was produced by `run_optimizer()` in `chapter06/optimizer_runtime.py`.

In [6]:
print(format_result(published_result(OPTIMIZER)))
print()
print(artifact_paths(OPTIMIZER))

OPTIMIZER: KNNFewShot
Validation Accuracy:   71.67% (43/60)
Locked Test Accuracy:  72.50% (58/80)
Same-Model Baseline:   53.75% (43/80)
Accuracy Uplift:       +18.75 percentage points
------------------------------------------------------------
Total Tokens:          137,377
  - Prompt:            117,253
  - Completion:        20,124
Optimization Cost:     $0.0000
Evaluation Cost:       $0.2380
Total Recorded Cost:   $0.2380
Optimization Time:     0.0s
Mean / p95 Latency:    1.830s / 2.837s

Saved artifacts:
- program snapshot: chapter06/results/expanded_notebooks/knn-few-shot/full/optimized_program.json
- prompt snapshot: chapter06/results/expanded_notebooks/knn-few-shot/full/learned_prompt.json
- chapter comparison: chapter06/CHAPTER_RESULTS.md


**Output printed in the book (page 141):**

```text
============================================================
OPTIMIZER: KNNFewShot
============================================================
Validation Accuracy:   71.67%
Locked Test Accuracy:  72.50% (58/80)
Same-Model Baseline:   53.75% (43/80)
Accuracy Uplift:       +18.75 percentage points
------------------------------------------------------------
Optimization Cost:     $0.0000
Evaluation Cost:       $0.2380
Optimization Time:     0.0s
Mean / p95 Latency:    1.830s / 2.837s
============================================================
```

These are the numbers of the saved result above; a live run is a new experiment, so its scores, cost, and timings will differ.

## Read the result

Compile cost is nearly zero, but retrieval and demonstration selection move into inference: every prediction first looks up its four nearest training examples. The saved program therefore stores no demonstrations, which is why the preview below reports `Demonstrations: 0`; KNNFewShot chooses them again for every input.

The saved program was compiled without the locked test: the optimizer saw training and validation data only, and the locked test was scored once afterwards.

> **Book vs. notebook:** The recorded 72.50% run additionally passed `max_bootstrapped_demos=0, max_labeled_demos=4`, so the four retrieved neighbors were used directly as labeled demonstrations with one model call per prediction. As printed, DSPy also bootstraps traces on the retrieved neighbors for every prediction, so a live run makes several model calls per prediction and costs more than the saved result.

The next cell previews the saved program.

In [7]:
print(learned_program_preview(OPTIMIZER))
print()
print("Saved program matches saved prompt:", verify_prompt_artifact(OPTIMIZER))

Predictor: detect.predict
Learned instruction (59 characters):
Decide whether the supplied passage was generated by an AI.

Demonstrations: 0

Saved program matches saved prompt: {'checked': True, 'predictors': 1, 'prompt_state_equal': True, 'mismatches': []}


## Apply the pattern

Adapt the listing above to your own DSPy program, metric, and frozen train/validation split. Keep the test set untouched until the optimizer returns, then report locked-test accuracy as `correct / test examples` so every optimizer is easy to compare. Use the baseline notebook (`quickstart-ai-detector.ipynb`) when you also need uplift.

All Chapter 6 results are summarized in `CHAPTER_RESULTS.md`. Scores, prompts, programs, and predictions are under `results/expanded_notebooks/` (GEPA: `results/gepa_light_standard/`). Fine-tuned model weights are generated locally and are not stored in the repository. No credentials are stored.